src/cpp/session/modules/SessionRAddins.cpp

```diff
@@ -0,0 +1,315 @@
+   
+private:
+   
+   static std::string constructKey(const std::string& package, const std::string& name)
+   {
+      return package + "::" + name;
+   }
+
+   std::map<std::string, AddinSpecification> addins_;
+};
+
+AddinRegistry& addinRegistry()
+{
+   static AddinRegistry instance;
+   return instance;
+}
+
+// TODO: This probably belongs in a separate module and could be exported
+std::vector<FilePath> getLibPaths()
+{
+   std::vector<std::string> libPathsString;
+   r::exec::RFunction rfLibPaths(".libPaths");
+   Error error = rfLibPaths.call(&libPathsString);
+   if (error)
+      LOG_ERROR(error);
+   
+   std::vector<FilePath> libPaths(libPathsString.size());
+   BOOST_FOREACH(const std::string& path, libPathsString)
+   {
+      libPaths.push_back(module_context::resolveAliasedPath(path));
+   }
+   
+   return libPaths;
+}
+
+std::map<std::string, std::string> parseAddinDcf(const std::string& contents)
+{
+   // read and parse the DCF file
+   std::map<std::string, std::string> fields;
+   std::string errMsg;
+   Error error = text::parseDcfFile(contents, true, &fields, &errMsg);
+   if (error)
+      LOG_ERROR(error);
+   
+   return fields;
+}
+
+void registerAddin(const std::string& pkgName,
+                   std::map<std::string, std::string>& fields)
+{
+   addinRegistry().add(pkgName, AddinSpecification(
+         fields["Name"],
+         pkgName,
+         fields["Title"],
+         fields["Description"],
+         fields["Binding"]));
+}
+
+void registerAddins(const std::string& pkgName, const FilePath& addinPath)
+{
+   static const boost::regex reSeparator("\\n{2,}");
+   
+   std::string contents;
+   Error error = core::readStringFromFile(addinPath, &contents, string_utils::LineEndingPosix);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   boost::sregex_token_iterator it(contents.begin(), contents.end(), reSeparator, -1);
+   boost::sregex_token_iterator end;
+   
+   for (; it != end; ++it)
+   {
+      std::map<std::string, std::string> fields = parseAddinDcf(*it);
+      registerAddin(pkgName, fields);
+   }
+}
+
+void discoverAddins(const std::vector<FilePath>& libPaths)
+{
+   BOOST_FOREACH(const FilePath& libPath, libPaths)
+   {
+      if (!libPath.exists())
+         continue;
+      
+      // get the children in this directory
+      std::vector<FilePath> pkgPaths;
+      Error error = libPath.children(&pkgPaths);
+      if (error)
+      {
+         LOG_ERROR(error);
+         continue;
+      }
+      
+      BOOST_FOREACH(const FilePath& pkgPath, pkgPaths)
+      {
+         // construct the addin path and check if it exists
+         FilePath addinPath = pkgPath.complete("rstudio/addins.dcf");
+         if (!addinPath.exists())
+            continue;
+         
+         std::string pkgName = pkgPath.filename();
+         registerAddins(pkgName, addinPath);
+      }
+   }
+}
+
+void onDeferredInit(bool newSession)
+{
+   discoverAddins(getLibPaths());
```

Comment:
Because of the unbounded nature of addin discovery I think you want to do this via the `scheduleIncrementalWork` API (that's how we do code indexing): https://github.com/rstudio/rstudio/blob/master/src/cpp/session/include/session/SessionModuleContext.hpp#L336-L350